tools/congestion-model/src/strategy/fancy_global_transaction_stop.rs

```diff
@@ -0,0 +1,288 @@
+use crate::strategy::QueueFactory;
+use crate::{GGas, QueueId, Receipt, ShardId, GAS_LIMIT};
+
+/// Fancy Global Transaction Stop implements the following ideas:
+/// * At any shard we have a dedicated outgoing queue for every other shard.
+/// * There are two types of congestion
+///   * The "incoming receipts" congestion
+///   * The "outgoing receipts" congestion
+/// * The incoming receipts congestion is dealt with by:
+///   * The shard, by processing incoming receipts before accepting new transactions.
+///   * The other shards, by slowing down the outflow of receipts.
+/// * The outgoing receipts congestion is dealt with by:
+///   * All shards, by stopping accepting new transactions
+///
+/// Deadlock prevention
+/// This strategy should never deadlock.
+/// For the sake of the argument let's assume there is a deadlock.
+/// * If incoming receipts is not empty those receipts will be processed and
+///   progress will be made.
+/// Assume incoming receipts is empty
+/// * If any outgoing receipts queue is non empty then some receipts will be
+///   forwarded and progress will be made.
+/// Assume outgoing receipts queue is empty
+/// * When outgoing receipts queues are empty new transactions will be included
+///   and progress will be made.
+///
+/// This can be improved with some small heuristics here and there. I'm not
+/// implementing it right now because I want to make sure that the core ideas
+/// work without help.
+/// * round robin full outflow allowance
+/// * filter out transactions aimed at congested shards
+pub struct FancyGlobalTransactionStop {
+    pub shard_id: Option<ShardId>,
+    pub all_shards: Vec<ShardId>,
+
+    // The soft limit on the incoming receipts queue attached gas. We begin to
+    // consider a shard congested when the incoming queue attached gas exceeds
+    // half of this value. As the incoming queue attached gas approaches the
+    // limit other shards will slow down the flow of receipts down to zero.
+    pub incoming_attached_gas_limit: GGas,
+    // The maximum on how much attached gas we are allowed to send to other
+    // shard in a single round. This is a per shard value. In practice the limit
+    // can be reduced based on the incoming receipts congestion in the receiving
+    // shard.
+    pub outgoing_attached_gas_limit: GGas,
+
+    // The queues for receipts going to other shards.
+    pub outgoing_receipts: BTreeMap<ShardId, QueueId>,
+
+    // How much attached gas was sent to other shards this round.
+    pub sent_outgoing_receipts_gas: BTreeMap<ShardId, GGas>,
+}
+
+struct CongestedShardsInfo {
+    // The attached gas of the incoming receipts queue.
+    // An indicator of the incoming receipts congestion.
+    incoming_receipts_attached_gas: GGas,
+
+    // The max of the attached gas of the outgoing receipts queues.
+    // An indicator of the outgoing receipts congestion.
+    // TODO perhaps sum would be better
+    max_outgoing_receipts_attached_gas: GGas,
+}
```

Comment:
I guess we can merge it as is and change it later if you think it would be better to use the sum?
Up to you.